File: sdk/python/src/agentspan/agents/cli_config.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""First-class CLI command execution configuration for agents.

Provides :class:`CliConfig` for declarative CLI tool attachment on
:class:`Agent`, a validation helper, and a factory function that
auto-creates a ``run_command`` tool.

Example::

    from agentspan.agents import Agent, CliConfig

    # Simple — just flip the flag
    agent = Agent(
        name="ops",
        model="openai/gpt-4o",
        cli_commands=True,
        cli_allowed_commands=["git", "gh", "curl"],
    )

    # Full control
    agent = Agent(
        name="ops",
        model="openai/gpt-4o",
        cli_config=CliConfig(
            allowed_commands=["git", "gh"],
            timeout=60,
            allow_shell=True,
        ),
    )
"""

from __future__ import annotations

import os
import shlex
import subprocess
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional


class TerminalToolError(RuntimeError):
    """Non-retryable tool failure.

    Causes the Conductor task to be marked FAILED_WITH_TERMINAL_ERROR
    instead of FAILED (which would trigger retries).
    """


@dataclass
class CliConfig:
    """Configuration for first-class CLI command execution on an Agent.

    Attributes:
        enabled: Whether CLI execution is active (default ``True``).
        allowed_commands: Command whitelist (e.g. ``["git", "gh"]``).
            Empty list means **no restrictions**.
        timeout: Maximum execution time in seconds (default ``30``).
        working_dir: Default working directory for commands.
        allow_shell: Config-level gate: can the LLM use ``shell=True``?
    """

    enabled: bool = True
    allowed_commands: List[str] = field(default_factory=list)
    timeout: int = 30
    working_dir: Optional[str] = None
    allow_shell: bool = False


# ── Validation ─────────────────────────────────────────────────────────


def _validate_cli_command(command: str, allowed_commands: List[str]) -> None:
    """Validate *command* against the whitelist.

    Strips path prefix (``/usr/bin/git`` → ``git``) before checking.
    Empty whitelist permits all commands.

    Raises:
        ValueError: If the command is not in the whitelist.
    """
    if not allowed_commands:
        return  # no restrictions
    base = os.path.basename(command)
    if base not in allowed_commands:
        raise ValueError(
            f"Command '{base}' is not allowed. "
            f"Allowed commands: {', '.join(sorted(allowed_commands))}"
        )


# ── Tool factory ───────────────────────────────────────────────────────


def _make_cli_tool(
    allowed_commands: List[str],
    timeout: int = 30,
    working_dir: Optional[str] = None,
    allow_shell: bool = False,
    agent_name: str = "",
) -> Any:
    """Create a ``@tool``-decorated ``run_command`` function.

    The returned function can be appended to ``Agent.tools`` directly.
    The tool name is prefixed with the agent name to avoid collisions
    when multiple agents define CLI tools with different allowed commands.
    """
    from agentspan.agents.tool import tool

    task_name = f"{agent_name}_run_command" if agent_name else "run_command"

    @tool(name=task_name)
    def run_command(
        command: str,
        args: list = [],
        cwd: str = "",
        shell: bool = False,
        context_key: str = "",
        context: Any = None,
        _allowed_commands: list = None,
        _allow_shell: bool = None,
        _timeout: int = None,
        _working_dir: str = None,
    ) -> dict:
        """Run a CLI command."""
        if not command or not isinstance(command, str):
            return {
                "status": "error",
                "stdout": "",
                "stderr": "No command provided.",
            }

        # Per-task overrides from server take precedence over closure defaults
        effective_allowed = _allowed_commands if _allowed_commands is not None else allowed_commands
        effective_allow_shell = _allow_shell if _allow_shell is not None else allow_shell
        effective_timeout = _timeout if _timeout is not None else timeout
        effective_working_dir = _working_dir if _working_dir is not None else working_dir

        # Validate against whitelist
        _validate_cli_command(command, effective_allowed)

        # Shell gate
        if shell and not effective_allow_shell:
            raise ValueError(
                "Shell mode is disabled for this agent. "
                "Do not set shell=True."
            )

        # Normalise args
        if args is None:
            args = []
        if not isinstance(args, list):
            args = [str(args)]

        # Resolve working directory
        effective_cwd = cwd if cwd else effective_working_dir

        try:
            if shell:
                # Build a safe shell command string
                cmd_str = command + " " + " ".join(
                    shlex.quote(str(a)) for a in args
                )
                result = subprocess.run(
                    cmd_str,
                    shell=True,
                    capture_output=True,
                    text=True,
                    timeout=effective_timeout,
                    cwd=effective_cwd or None,
                )
            else:
                result = subprocess.run(
                    [command] + [str(a) for a in args],
                    capture_output=True,
                    text=True,
                    timeout=effective_timeout,
                    cwd=effective_cwd or None,
                )

            if result.returncode == 0:
                if context_key and context is not None and hasattr(context, "state"):
                    # Prefer stdout; fall back to stderr (e.g. git clone outputs to stderr)
                    value = result.stdout.strip() or result.stderr.strip()
                    if value:
                        context.state[context_key] = value
                return {
                    "status": "success",
                    "exit_code": 0,
                    "stdout": result.stdout,
                    "stderr": result.stderr,
                }
            else:
                return {
                    "status": "error",
                    "exit_code": result.returncode,
                    "stdout": result.stdout,
                    "stderr": result.stderr,
                }

        except subprocess.TimeoutExpired:
            raise TerminalToolError(
                f"Command timed out after {effective_timeout}s"
            )
        except FileNotFoundError:
            raise TerminalToolError(f"Command not found: {command}")

    # Build dynamic description
    desc = (
        "Run a CLI command directly. "
        f"Timeout: {timeout}s."
    )
    if allowed_commands:
        desc += f" Allowed commands: {', '.join(sorted(allowed_commands))}."
    if not allow_shell:
        desc += " Shell mode is disabled — do not set shell=True."
    desc += (
        " If you need to save a command's output for later pipeline steps, set context_key."
        " Well-known keys: repo, branch, working_dir, issue_number, pr_url, commit_sha."
    )
    run_command._tool_def.description = desc
    run_command._tool_def.tool_type = "cli"
    run_command._tool_def.config = {"allowedCommands": list(allowed_commands)}

    return run_command

File: sdk/typescript/src/cli-config.ts
/**
 * First-class CLI command execution configuration for agents.
 *
 * Provides {@link CliConfigOptions} for declarative CLI tool attachment on
 * {@link Agent}, a validation helper, and a factory function that
 * auto-creates a `run_command` tool.
 *
 * Example:
 *
 *   import { Agent } from 'agentspan';
 *
 *   // Simple — just flip the flag
 *   const agent = new Agent({
 *     name: 'ops',
 *     model: 'openai/gpt-4o',
 *     cliCommands: true,
 *     cliAllowedCommands: ['git', 'gh', 'curl'],
 *   });
 *
 *   // Full control
 *   import { CliConfigOptions } from 'agentspan';
 *
 *   const agent = new Agent({
 *     name: 'ops',
 *     model: 'openai/gpt-4o',
 *     cliConfig: {
 *       enabled: true,
 *       allowedCommands: ['git', 'gh'],
 *       timeout: 60,
 *       allowShell: true,
 *     },
 *   });
 */

import { execSync } from 'child_process';
import { TerminalToolError } from './errors.js';
import type { ToolDef } from './types.js';

// ── CliConfigOptions ──────────────────────────────────────

/**
 * Configuration for first-class CLI command execution on an Agent.
 *
 * This is the *options* interface used for constructing agents.
 * The existing `CliConfig` type in types.ts is the wire/serialization format.
 */
export interface CliConfigOptions {
  /** Whether CLI execution is active (default true). */
  enabled?: boolean;
  /** Command whitelist (e.g. ['git', 'gh']). Empty means no restrictions. */
  allowedCommands?: string[];
  /** Maximum execution time in seconds (default 30). */
  timeout?: number;
  /** Default working directory for commands. */
  workingDir?: string;
  /** Config-level gate: can the LLM use shell mode? */
  allowShell?: boolean;
}

// ── Validation ────────────────────────────────────────────

/**
 * Validate a command against the whitelist.
 * Strips path prefix (/usr/bin/git -> git) before checking.
 * Empty whitelist permits all commands.
 */
function validateCliCommand(command: string, allowedCommands: string[]): void {
  if (!allowedCommands || allowedCommands.length === 0) {
    return; // no restrictions
  }
  // Strip path prefix
  const parts = command.split('/');
  const base = parts[parts.length - 1];
  if (!allowedCommands.includes(base)) {
    throw new Error(
      `Command '${base}' is not allowed. ` +
        `Allowed commands: ${[...allowedCommands].sort().join(', ')}`,
    );
  }
}

// ── Tool factory ──────────────────────────────────────────

/**
 * Create a ToolDef for CLI command execution.
 *
 * The returned ToolDef can be appended to Agent.tools directly.
 * The tool name is prefixed with the agent name to avoid collisions
 * when multiple agents define CLI tools with different allowed commands.
 */
export function makeCliTool(
  config: CliConfigOptions,
  agentName: string,
): ToolDef {
  const allowedCommands = config.allowedCommands ?? [];
  const timeout = config.timeout ?? 30;
  const workingDir = config.workingDir;
  const allowShell = config.allowShell ?? false;
  const taskName = agentName ? `${agentName}_run_command` : 'run_command';

  // Build dynamic description
  let desc = `Run a CLI command directly. Timeout: ${timeout}s.`;
  if (allowedCommands.length > 0) {
    desc += ` Allowed commands: ${[...allowedCommands].sort().join(', ')}.`;
  }
  if (!allowShell) {
    desc += ' Shell mode is disabled — do not set shell=true.';
  }

  return {
    name: taskName,
    description: desc,
    inputSchema: {
      type: 'object',
      properties: {
        command: { type: 'string', description: 'The CLI command to execute' },
        args: {
          type: 'array',
          items: { type: 'string' },
          description: 'Command arguments',
        },
        cwd: {
          type: 'string',
          description: 'Working directory for the command',
        },
        shell: {
          type: 'boolean',
          description: 'Whether to run via shell',
        },
      },
      required: ['command'],
    },
    toolType: 'worker',
    config: { allowedCommands: [...allowedCommands] },
    func: async (args: Record<string, unknown>) => {
      const command = args.command as string;
      if (!command || typeof command !== 'string') {
        return {
          status: 'error',
          stdout: '',
          stderr: 'No command provided.',
        };
      }

      // Validate against whitelist
      validateCliCommand(command, allowedCommands);

      // Shell gate
      const useShell = args.shell === true;
      if (useShell && !allowShell) {
        throw new Error(
          'Shell mode is disabled for this agent. Do not set shell=true.',
        );
      }

      // Normalise args
      let cmdArgs = (args.args as string[]) ?? [];
      if (!Array.isArray(cmdArgs)) {
        cmdArgs = [String(cmdArgs)];
      }

      // Resolve working directory
      const effectiveCwd =
        (args.cwd as string) || workingDir || undefined;

      try {
        let cmdStr: string;
        if (useShell) {
          cmdStr =
            command +
            ' ' +
            cmdArgs.map((a) => JSON.stringify(String(a))).join(' ');
        } else {
          const parts = [command, ...cmdArgs.map(String)];
          cmdStr = parts.map((p) => JSON.stringify(p)).join(' ');
        }

        const output = execSync(cmdStr, {
          timeout: timeout * 1000,
          encoding: 'utf-8',
          cwd: effectiveCwd,
          shell: useShell ? '/bin/sh' : undefined,
          stdio: ['pipe', 'pipe', 'pipe'],
        });

        return {
          status: 'success',
          exit_code: 0,
          stdout: output,
          stderr: '',
        };
      } catch (err: unknown) {
        const execErr = err as {
          status?: number | null;
          killed?: boolean;
          stdout?: string;
          stderr?: string;
          signal?: string;
          message?: string;
        };

        if (execErr.killed === true || execErr.signal === 'SIGTERM') {
          throw new TerminalToolError(`Command timed out after ${timeout}s`);
        }

        if (
          execErr.message &&
          execErr.message.includes('ENOENT')
        ) {
          throw new TerminalToolError(`Command not found: ${command}`);
        }

        const stdout =
          typeof execErr.stdout === 'string' ? execErr.stdout : '';
        const stderr =
          typeof execErr.stderr === 'string'
            ? execErr.stderr
            : String(err);
        const exitCode = execErr.status ?? 1;

        return {
          status: 'error',
          exit_code: exitCode,
          stdout,
          stderr,
        };
      }
    },
  };
}
